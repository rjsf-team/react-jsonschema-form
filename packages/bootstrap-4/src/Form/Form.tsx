import { withTheme, FormProps } from "@rjsf/core";

import Theme from "../Theme";
import { StatelessComponent } from "react";

const Form:
  | React.ComponentClass<FormProps<any>>
  | StatelessComponent<FormProps<any>> = withTheme(Theme);

export default Form;
