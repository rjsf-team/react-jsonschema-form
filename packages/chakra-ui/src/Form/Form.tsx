import { withTheme, FormProps } from "@rjsf/core";

import Theme from "../Theme";
import { ComponentClass, FunctionComponent } from "react";

const Form:
  | ComponentClass<FormProps<any>>
  | FunctionComponent<FormProps<any>> = withTheme(Theme);

export default Form;
