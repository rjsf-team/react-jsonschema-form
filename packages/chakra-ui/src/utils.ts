import { ChakraProps, shouldForwardProp } from '@chakra-ui/react';
import { UiSchema } from '@rjsf/utils';

export interface ChakraUiSchema extends Omit<UiSchema, 'ui:options'> {
  'ui:options'?: ChakraUiOptions;
}

type ChakraUiOptions = UiSchema['ui:options'] & { chakra?: ChakraProps };

interface GetChakraProps {
  uiSchema?: ChakraUiSchema;
}

export function getChakra({ uiSchema = {} }: GetChakraProps): ChakraProps {
  const chakraProps = (uiSchema['ui:options'] && uiSchema['ui:options'].chakra) || {};

  Object.keys(chakraProps).forEach((key) => {
    /**
     * Leveraging `shouldForwardProp` to remove props
     *
     * This is a utility function that's used in `@chakra-ui/react`'s factory function.
     * Normally, it prevents ChakraProps from being passed to the DOM.
     * In this case we just want to delete the unknown props. So we flip the boolean.
     */
    if (shouldForwardProp(key)) {
      delete (chakraProps as any)[key];
    }
  });

  return chakraProps;
}
