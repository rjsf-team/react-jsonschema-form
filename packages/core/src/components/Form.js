import React, { Component } from "react";
import PropTypes from "prop-types";
import _pick from "lodash/pick";
import _get from "lodash/get";
import _isEmpty from "lodash/isEmpty";

import { default as DefaultErrorList } from "./ErrorList";
import {
  getDefaultFormState,
  retrieveSchema,
  shouldRender,
  toIdSchema,
  getDefaultRegistry,
  deepEquals,
  toPathSchema,
  isObject,
} from "../utils";
import validateFormData, { toErrorList } from "../validate";
import { mergeObjects } from "../utils";

export default class Form extends Component {
  static defaultProps = {
    uiSchema: {},
    noValidate: false,
    liveValidate: false,
    disabled: false,
    readonly: false,
    noHtml5Validate: false,
    ErrorList: DefaultErrorList,
    omitExtraData: false,
  };

  constructor(props) {
    super(props);
    this.state = this.getStateFromProps(props, props.formData);
    if (
      this.props.onChange &&
      !deepEquals(this.state.formData, this.props.formData)
    ) {
      this.props.onChange(this.state);
    }
    this.formElement = null;
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    const nextState = this.getStateFromProps(nextProps, nextProps.formData);
    if (
      !deepEquals(nextState.formData, nextProps.formData) &&
      !deepEquals(nextState.formData, this.state.formData) &&
      this.props.onChange
    ) {
      this.props.onChange(nextState);
    }
    this.setState(nextState);
  }

  getStateFromProps(props, inputFormData) {
    const state = this.state || {};
    const schema = "schema" in props ? props.schema : this.props.schema;
    const uiSchema = "uiSchema" in props ? props.uiSchema : this.props.uiSchema;
    const edit = typeof inputFormData !== "undefined";
    const liveValidate =
      "liveValidate" in props ? props.liveValidate : this.props.liveValidate;
    const mustValidate = edit && !props.noValidate && liveValidate;
    const rootSchema = schema;
    const formData = getDefaultFormState(schema, inputFormData, rootSchema);
    const retrievedSchema = retrieveSchema(schema, rootSchema, formData);
    const customFormats = props.customFormats;
    const additionalMetaSchemas = props.additionalMetaSchemas;

    const getCurrentErrors = () => {
      if (props.noValidate) {
        return { errors: [], errorSchema: {} };
      } else if (!props.liveValidate) {
        return {
          errors: state.schemaValidationErrors || [],
          errorSchema: state.schemaValidationErrorSchema || {},
        };
      }
      return {
        errors: state.errors || [],
        errorSchema: state.errorSchema || {},
      };
    };

    let errors,
      errorSchema,
      schemaValidationErrors,
      schemaValidationErrorSchema;
    if (mustValidate) {
      const schemaValidation = this.validate(
        formData,
        schema,
        additionalMetaSchemas,
        customFormats
      );
      errors = schemaValidation.errors;
      errorSchema = schemaValidation.errorSchema;
      schemaValidationErrors = errors;
      schemaValidationErrorSchema = errorSchema;
    } else {
      const currentErrors = getCurrentErrors();
      errors = currentErrors.errors;
      errorSchema = currentErrors.errorSchema;
      schemaValidationErrors = state.schemaValidationErrors;
      schemaValidationErrorSchema = state.schemaValidationErrorSchema;
    }
    if (props.extraErrors) {
      errorSchema = mergeObjects(
        errorSchema,
        props.extraErrors,
        !!"concat arrays"
      );
      errors = toErrorList(errorSchema);
    }
    const idSchema = toIdSchema(
      retrievedSchema,
      uiSchema["ui:rootFieldId"],
      rootSchema,
      formData,
      props.idPrefix
    );
    const nextState = {
      schema,
      uiSchema,
      idSchema,
      formData,
      edit,
      errors,
      errorSchema,
      additionalMetaSchemas,
    };
    if (schemaValidationErrors) {
      nextState.schemaValidationErrors = schemaValidationErrors;
      nextState.schemaValidationErrorSchema = schemaValidationErrorSchema;
    }
    return nextState;
  }

  shouldComponentUpdate(nextProps, nextState) {
    return shouldRender(this, nextProps, nextState);
  }

  validate(
    formData,
    schema = this.props.schema,
    additionalMetaSchemas = this.props.additionalMetaSchemas,
    customFormats = this.props.customFormats
  ) {
    const { validate, transformErrors } = this.props;
    const { rootSchema } = this.getRegistry();
    const resolvedSchema = retrieveSchema(schema, rootSchema, formData);
    return validateFormData(
      formData,
      resolvedSchema,
      validate,
      transformErrors,
      additionalMetaSchemas,
      customFormats
    );
  }

  renderErrors() {
    const { errors, errorSchema, schema, uiSchema } = this.state;
    const { ErrorList, showErrorList, formContext } = this.props;

    if (errors.length && showErrorList != false) {
      return (
        <ErrorList
          errors={errors}
          errorSchema={errorSchema}
          schema={schema}
          uiSchema={uiSchema}
          formContext={formContext}
        />
      );
    }
    return null;
  }

  getUsedFormData = (formData, fields) => {
    //for the case of a single input form
    if (fields.length === 0 && typeof formData !== "object") {
      return formData;
    }

    let data = _pick(formData, fields);
    if (Array.isArray(formData)) {
      return Object.keys(data).map(key => data[key]);
    }

    return data;
  };

  getFieldNames = (pathSchema, formData) => {
    const getAllPaths = (_obj, acc = [], paths = [""]) => {
      Object.keys(_obj).forEach(key => {
        if (typeof _obj[key] === "object") {
          let newPaths = paths.map(path => `${path}.${key}`);
          // If an object is marked with additionalProperties, all its keys are valid
          if (_obj[key].__rjsf_additionalProperties && _obj[key].$name !== "") {
            acc.push(_obj[key].$name);
          } else {
            getAllPaths(_obj[key], acc, newPaths);
          }
        } else if (key === "$name" && _obj[key] !== "") {
          paths.forEach(path => {
            path = path.replace(/^\./, "");
            const formValue = _get(formData, path);
            // adds path to fieldNames if it points to a value
            // or an empty object/array
            if (typeof formValue !== "object" || _isEmpty(formValue)) {
              acc.push(path);
            }
          });
        }
      });
      return acc;
    };

    return getAllPaths(pathSchema);
  };

  onChange = (formData, newErrorSchema) => {
    if (isObject(formData) || Array.isArray(formData)) {
      const newState = this.getStateFromProps(this.props, formData);
      formData = newState.formData;
    }
    const mustValidate = !this.props.noValidate && this.props.liveValidate;
    let state = { formData };
    let newFormData = formData;

    if (this.props.omitExtraData === true && this.props.liveOmit === true) {
      const retrievedSchema = retrieveSchema(
        this.state.schema,
        this.state.schema,
        formData
      );
      const pathSchema = toPathSchema(
        retrievedSchema,
        "",
        this.state.schema,
        formData
      );

      const fieldNames = this.getFieldNames(pathSchema, formData);

      newFormData = this.getUsedFormData(formData, fieldNames);
      state = {
        formData: newFormData,
      };
    }

    if (mustValidate) {
      let schemaValidation = this.validate(newFormData);
      let errors = schemaValidation.errors;
      let errorSchema = schemaValidation.errorSchema;
      const schemaValidationErrors = errors;
      const schemaValidationErrorSchema = errorSchema;
      if (this.props.extraErrors) {
        errorSchema = mergeObjects(
          errorSchema,
          this.props.extraErrors,
          !!"concat arrays"
        );
        errors = toErrorList(errorSchema);
      }
      state = {
        formData: newFormData,
        errors,
        errorSchema,
        schemaValidationErrors,
        schemaValidationErrorSchema,
      };
    } else if (!this.props.noValidate && newErrorSchema) {
      const errorSchema = this.props.extraErrors
        ? mergeObjects(
            newErrorSchema,
            this.props.extraErrors,
            !!"concat arrays"
          )
        : newErrorSchema;
      state = {
        formData: newFormData,
        errorSchema: errorSchema,
        errors: toErrorList(errorSchema),
      };
    }
    this.setState(
      state,
      () => this.props.onChange && this.props.onChange(this.state)
    );
  };

  onBlur = (...args) => {
    if (this.props.onBlur) {
      this.props.onBlur(...args);
    }
  };

  onFocus = (...args) => {
    if (this.props.onFocus) {
      this.props.onFocus(...args);
    }
  };

  onSubmit = event => {
    event.preventDefault();
    if (event.target !== event.currentTarget) {
      return;
    }

    event.persist();
    let newFormData = this.state.formData;

    if (this.props.omitExtraData === true) {
      const retrievedSchema = retrieveSchema(
        this.state.schema,
        this.state.schema,
        newFormData
      );
      const pathSchema = toPathSchema(
        retrievedSchema,
        "",
        this.state.schema,
        newFormData
      );

      const fieldNames = this.getFieldNames(pathSchema, newFormData);

      newFormData = this.getUsedFormData(newFormData, fieldNames);
    }

    if (!this.props.noValidate) {
      let schemaValidation = this.validate(newFormData);
      let errors = schemaValidation.errors;
      let errorSchema = schemaValidation.errorSchema;
      const schemaValidationErrors = errors;
      const schemaValidationErrorSchema = errorSchema;
      if (Object.keys(errors).length > 0) {
        if (this.props.extraErrors) {
          errorSchema = mergeObjects(
            errorSchema,
            this.props.extraErrors,
            !!"concat arrays"
          );
          errors = toErrorList(errorSchema);
        }
        this.setState(
          {
            errors,
            errorSchema,
            schemaValidationErrors,
            schemaValidationErrorSchema,
          },
          () => {
            if (this.props.onError) {
              this.props.onError(errors);
            } else {
              console.error("Form validation failed", errors);
            }
          }
        );
        return;
      }
    }

    // There are no errors generated through schema validation.
    // Check for user provided errors and update state accordingly.
    let errorSchema;
    let errors;
    if (this.props.extraErrors) {
      errorSchema = this.props.extraErrors;
      errors = toErrorList(errorSchema);
    } else {
      errorSchema = {};
      errors = [];
    }

    this.setState(
      {
        formData: newFormData,
        errors: errors,
        errorSchema: errorSchema,
        schemaValidationErrors: [],
        schemaValidationErrorSchema: {},
      },
      () => {
        if (this.props.onSubmit) {
          this.props.onSubmit(
            { ...this.state, formData: newFormData, status: "submitted" },
            event
          );
        }
      }
    );
  };

  getRegistry() {
    // For BC, accept passed SchemaField and TitleField props and pass them to
    // the "fields" registry one.
    const { fields, widgets } = getDefaultRegistry();
    return {
      fields: { ...fields, ...this.props.fields },
      widgets: { ...widgets, ...this.props.widgets },
      ArrayFieldTemplate: this.props.ArrayFieldTemplate,
      ObjectFieldTemplate: this.props.ObjectFieldTemplate,
      FieldTemplate: this.props.FieldTemplate,
      definitions: this.props.schema.definitions || {},
      rootSchema: this.props.schema,
      formContext: this.props.formContext || {},
    };
  }

  submit() {
    if (this.formElement) {
      this.formElement.dispatchEvent(
        new CustomEvent("submit", {
          cancelable: true,
        })
      );
    }
  }

  render() {
    const {
      children,
      id,
      idPrefix,
      className,
      tagName,
      name,
      method,
      target,
      action,
      autocomplete: deprecatedAutocomplete,
      autoComplete: currentAutoComplete,
      enctype,
      acceptcharset,
      noHtml5Validate,
      disabled,
      readonly,
      formContext,
    } = this.props;

    const { schema, uiSchema, formData, errorSchema, idSchema } = this.state;
    const registry = this.getRegistry();
    const _SchemaField = registry.fields.SchemaField;
    const FormTag = tagName ? tagName : "form";
    if (deprecatedAutocomplete) {
      console.warn(
        "Using autocomplete property of Form is deprecated, use autoComplete instead."
      );
    }
    const autoComplete = currentAutoComplete
      ? currentAutoComplete
      : deprecatedAutocomplete;

    return (
      <FormTag
        className={className ? className : "rjsf"}
        id={id}
        name={name}
        method={method}
        target={target}
        action={action}
        autoComplete={autoComplete}
        encType={enctype}
        acceptCharset={acceptcharset}
        noValidate={noHtml5Validate}
        onSubmit={this.onSubmit}
        ref={form => {
          this.formElement = form;
        }}>
        {this.renderErrors()}
        <_SchemaField
          schema={schema}
          uiSchema={uiSchema}
          errorSchema={errorSchema}
          idSchema={idSchema}
          idPrefix={idPrefix}
          formContext={formContext}
          formData={formData}
          onChange={this.onChange}
          onBlur={this.onBlur}
          onFocus={this.onFocus}
          registry={registry}
          disabled={disabled}
          readonly={readonly}
        />
        {children ? (
          children
        ) : (
          <div>
            <button type="submit" className="btn btn-info">
              Submit
            </button>
          </div>
        )}
      </FormTag>
    );
  }
}

if (process.env.NODE_ENV !== "production") {
  Form.propTypes = {
    schema: PropTypes.object.isRequired,
    uiSchema: PropTypes.object,
    formData: PropTypes.any,
    disabled: PropTypes.bool,
    readonly: PropTypes.bool,
    widgets: PropTypes.objectOf(
      PropTypes.oneOfType([PropTypes.func, PropTypes.object])
    ),
    fields: PropTypes.objectOf(PropTypes.elementType),
    ArrayFieldTemplate: PropTypes.elementType,
    ObjectFieldTemplate: PropTypes.elementType,
    FieldTemplate: PropTypes.elementType,
    ErrorList: PropTypes.func,
    onChange: PropTypes.func,
    onError: PropTypes.func,
    showErrorList: PropTypes.bool,
    onSubmit: PropTypes.func,
    id: PropTypes.string,
    className: PropTypes.string,
    tagName: PropTypes.elementType,
    name: PropTypes.string,
    method: PropTypes.string,
    target: PropTypes.string,
    action: PropTypes.string,
    autocomplete: PropTypes.string,
    autoComplete: PropTypes.string,
    enctype: PropTypes.string,
    acceptcharset: PropTypes.string,
    noValidate: PropTypes.bool,
    noHtml5Validate: PropTypes.bool,
    liveValidate: PropTypes.bool,
    validate: PropTypes.func,
    transformErrors: PropTypes.func,
    formContext: PropTypes.object,
    customFormats: PropTypes.object,
    additionalMetaSchemas: PropTypes.arrayOf(PropTypes.object),
    omitExtraData: PropTypes.bool,
    extraErrors: PropTypes.object,
  };
}
