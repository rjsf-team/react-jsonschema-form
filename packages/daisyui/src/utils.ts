import { UiSchema } from '@rjsf/utils';

import { DaisyProps } from './types/DaisyProps';

export interface DaisyUiSchema extends Omit<UiSchema, 'ui:options'> {
  'ui:options'?: DaisyUiOptions;
}

type DaisyUiOptions = UiSchema['ui:options'] & { daisy?: DaisyProps };

interface GetDaisyProps {
  uiSchema?: DaisyUiSchema;
}

export function getDaisy({ uiSchema = {} }: GetDaisyProps): DaisyProps {
  const daisyProps = (uiSchema['ui:options'] && uiSchema['ui:options'].daisy) || {};
  return daisyProps;
}
