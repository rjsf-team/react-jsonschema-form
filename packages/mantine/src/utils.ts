import { UIOptionsType } from '@rjsf/utils';

const uiOptionsKeys: Array<keyof UIOptionsType> = [
  'emptyValue',
  'classNames',
  'title',
  'help',
  'autocomplete',
  'disabled',
  'enumDisabled',
  'hideError',
  'readonly',
  'order',
  'filePreview',
  'inline',
  'inputType',
  'submitButtonOptions',
  'widget',
  'enumNames',
  'addable',
  'copyable',
  'orderable',
  'removable',
  'duplicateKeySuffixSeparator',
  'enumOptions',
  'enableMarkdownInDescription',
];

export function cleanupOptions<T extends object>(options: T): Omit<T, keyof UIOptionsType> {
  const result = {} as T;
  for (const key in options) {
    if (!uiOptionsKeys.includes(key as keyof UIOptionsType)) {
      result[key] = options[key];
    }
  }
  return result as Omit<T, keyof UIOptionsType>;
}
