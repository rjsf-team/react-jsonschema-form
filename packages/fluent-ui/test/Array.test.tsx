import React from "react";
import { RJSFSchema, ErrorSchema } from "@rjsf/utils";
import validator from "@rjsf/validator-ajv6";
import renderer from "react-test-renderer";

import Form from "../src/index";

describe("array fields", () => {
  test("array", () => {
    const schema: RJSFSchema = {
      type: "array",
      items: {
        type: "string",
      },
    };
    const tree = renderer
      .create(<Form schema={schema} validator={validator} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("fixed array", () => {
    const schema: RJSFSchema = {
      type: "array",
      items: [
        {
          type: "string",
        },
        {
          type: "number",
        },
      ],
    };
    const tree = renderer
      .create(<Form schema={schema} validator={validator} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("checkboxes", () => {
    const schema: RJSFSchema = {
      type: "array",
      items: {
        type: "string",
        enum: ["a", "b", "c"],
      },
      uniqueItems: true,
    };
    const tree = renderer
      .create(<Form schema={schema} validator={validator} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("array icons", () => {
    const schema: RJSFSchema = {
      type: "array",
      items: {
        type: "string",
      },
    };
    const tree = renderer
      .create(
        <Form schema={schema} validator={validator} formData={["a", "b"]} />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("no errors", () => {
    const schema: RJSFSchema = {
      type: "object",
      properties: {
        name: {
          type: "string",
        },
      },
    };
    const tree = renderer
      .create(<Form schema={schema} validator={validator} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("empty errors array", () => {
    const schema: RJSFSchema = {
      type: "object",
      properties: {
        name: {
          type: "string",
        },
      },
    };
    const errors: any[] = [];
    const extraErrors = {
      name: { __errors: errors },
    } as unknown as ErrorSchema;
    const tree = renderer
      .create(
        <Form schema={schema} validator={validator} extraErrors={extraErrors} />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
