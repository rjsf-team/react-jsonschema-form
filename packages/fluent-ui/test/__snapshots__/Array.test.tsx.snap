// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array fields array 1`] = `
<form
  className="rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="ms-Grid-col ms-sm12  field field-array"
    style={
      Object {
        "display": undefined,
        "marginBottom": 15,
      }
    }
  >
    <span
      style={
        Object {
          "float": "right",
        }
      }
    >
      <button
        className="ms-Button ms-Button--commandBar array-item-add root-41"
        data-is-focusable={true}
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        style={
          Object {
            "height": "32px",
          }
        }
        type="button"
      >
        <span
          className="ms-Button-flexContainer flexContainer-42"
          data-automationid="splitbuttonprimary"
        >
          <i
            aria-hidden={true}
            className="ms-Icon root-37 css-49 ms-Button-icon icon-44"
            data-icon-name="Add"
            role="presentation"
            style={
              Object {
                "fontFamily": "\\"FabricMDL2Icons\\"",
              }
            }
          >
            
          </i>
          <span
            className="ms-Button-textContainer textContainer-43"
          >
            <span
              className="ms-Button-label label-45"
              id="id__1"
            >
              Add item
            </span>
          </span>
        </span>
      </button>
    </span>
  </div>
  <div>
    <br />
    <div
      className="ms-Grid-col ms-sm12"
    >
      <button
        className="ms-Button ms-Button--primary root-50"
        data-is-focusable={true}
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        type="submit"
      >
        <span
          className="ms-Button-flexContainer flexContainer-42"
          data-automationid="splitbuttonprimary"
        >
          <span
            className="ms-Button-textContainer textContainer-43"
          >
            <span
              className="ms-Button-label label-52"
              id="id__4"
            >
              Submit
            </span>
          </span>
        </span>
      </button>
    </div>
  </div>
</form>
`;

exports[`array fields checkboxes 1`] = `
<form
  className="rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="ms-Grid-col ms-sm12  field field-array"
    style={
      Object {
        "display": undefined,
        "marginBottom": 15,
      }
    }
  >
    <label
      className="ms-Label root-69"
    />
    <div
      className="ms-Dropdown-container"
    >
      <div
        aria-disabled={false}
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-required={false}
        className="ms-Dropdown dropdown-70"
        data-is-focusable={true}
        id="Dropdown34"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          aria-atomic={true}
          aria-invalid={false}
          aria-live="polite"
          className="ms-Dropdown-title ms-Dropdown-titleIsPlaceHolder title-71"
          id="Dropdown34-option"
        />
        <span
          className="ms-Dropdown-caretDownWrapper caretDownWrapper-72"
        >
          <i
            aria-hidden={true}
            className="ms-Dropdown-caretDown caretDown-86"
            data-icon-name="ChevronDown"
          >
            
          </i>
        </span>
      </div>
    </div>
  </div>
  <div>
    <br />
    <div
      className="ms-Grid-col ms-sm12"
    >
      <button
        className="ms-Button ms-Button--primary root-50"
        data-is-focusable={true}
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        type="submit"
      >
        <span
          className="ms-Button-flexContainer flexContainer-42"
          data-automationid="splitbuttonprimary"
        >
          <span
            className="ms-Button-textContainer textContainer-43"
          >
            <span
              className="ms-Button-label label-52"
              id="id__35"
            >
              Submit
            </span>
          </span>
        </span>
      </button>
    </div>
  </div>
</form>
`;

exports[`array fields fixed array 1`] = `
<form
  className="rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="ms-Grid-col ms-sm12  field field-array"
    style={
      Object {
        "display": undefined,
        "marginBottom": 15,
      }
    }
  >
    <div
      className="ms-Grid"
      dir="ltr"
    >
      <div
        className="ms-Grid-row"
      >
        <div
          className="ms-Grid-col ms-sm6 ms-md8 ms-lg9"
        >
          <div
            className="ms-Grid-row"
          >
            <div
              className="ms-Grid-col ms-sm12  field field-string"
              style={
                Object {
                  "display": undefined,
                  "marginBottom": 15,
                }
              }
            >
              <div
                className="ms-TextField is-required root-55"
              >
                <div
                  className="ms-TextField-wrapper"
                >
                  <div
                    className="ms-TextField-fieldGroup fieldGroup-56"
                  >
                    <input
                      aria-invalid={false}
                      autoFocus={false}
                      className="ms-TextField-field field-57"
                      disabled={false}
                      id="root_0"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      onInput={[Function]}
                      placeholder=""
                      readOnly={false}
                      required={true}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="ms-Grid-col ms-sm6 ms-md4 ms-lg3"
          style={
            Object {
              "textAlign": "right",
            }
          }
        >
          <button
            aria-disabled={true}
            className="ms-Button ms-Button--icon is-disabled root-63"
            data-is-focusable={false}
            disabled={true}
            onClick={[Function]}
            onKeyDown={[Function]}
            onKeyPress={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            type="button"
          >
            <span
              className="ms-Button-flexContainer flexContainer-42"
              data-automationid="splitbuttonprimary"
            >
              <i
                aria-hidden={true}
                className="ms-Icon root-37 css-67 ms-Button-icon icon-64"
                data-icon-name="Up"
                role="presentation"
                style={
                  Object {
                    "fontFamily": "\\"FabricMDL2Icons-0\\"",
                  }
                }
              >
                
              </i>
            </span>
          </button>
          <button
            aria-disabled={true}
            className="ms-Button ms-Button--icon is-disabled root-63"
            data-is-focusable={false}
            disabled={true}
            onClick={[Function]}
            onKeyDown={[Function]}
            onKeyPress={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            type="button"
          >
            <span
              className="ms-Button-flexContainer flexContainer-42"
              data-automationid="splitbuttonprimary"
            >
              <i
                aria-hidden={true}
                className="ms-Icon root-37 css-67 ms-Button-icon icon-64"
                data-icon-name="Down"
                role="presentation"
                style={
                  Object {
                    "fontFamily": "\\"FabricMDL2Icons-0\\"",
                  }
                }
              >
                
              </i>
            </span>
          </button>
          <button
            className="ms-Button ms-Button--icon root-68"
            data-is-focusable={true}
            onClick={[Function]}
            onKeyDown={[Function]}
            onKeyPress={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            type="button"
          >
            <span
              className="ms-Button-flexContainer flexContainer-42"
              data-automationid="splitbuttonprimary"
            >
              <i
                aria-hidden={true}
                className="ms-Icon root-37 css-49 ms-Button-icon icon-51"
                data-icon-name="Delete"
                role="presentation"
                style={
                  Object {
                    "fontFamily": "\\"FabricMDL2Icons\\"",
                  }
                }
              >
                
              </i>
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      className="ms-Grid"
      dir="ltr"
    >
      <div
        className="ms-Grid-row"
      >
        <div
          className="ms-Grid-col ms-sm6 ms-md8 ms-lg9"
        >
          <div
            className="ms-Grid-row"
          >
            <div
              className="ms-Grid-col ms-sm12  field field-number"
              style={
                Object {
                  "display": undefined,
                  "marginBottom": 15,
                }
              }
            >
              <div
                className="ms-TextField is-required root-55"
              >
                <div
                  className="ms-TextField-wrapper"
                >
                  <div
                    className="ms-TextField-fieldGroup fieldGroup-56"
                  >
                    <input
                      aria-invalid={false}
                      autoFocus={false}
                      className="ms-TextField-field field-57"
                      disabled={false}
                      id="root_1"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      onInput={[Function]}
                      placeholder=""
                      readOnly={false}
                      required={true}
                      type="number"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="ms-Grid-col ms-sm6 ms-md4 ms-lg3"
          style={
            Object {
              "textAlign": "right",
            }
          }
        >
          <button
            aria-disabled={true}
            className="ms-Button ms-Button--icon is-disabled root-63"
            data-is-focusable={false}
            disabled={true}
            onClick={[Function]}
            onKeyDown={[Function]}
            onKeyPress={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            type="button"
          >
            <span
              className="ms-Button-flexContainer flexContainer-42"
              data-automationid="splitbuttonprimary"
            >
              <i
                aria-hidden={true}
                className="ms-Icon root-37 css-67 ms-Button-icon icon-64"
                data-icon-name="Up"
                role="presentation"
                style={
                  Object {
                    "fontFamily": "\\"FabricMDL2Icons-0\\"",
                  }
                }
              >
                
              </i>
            </span>
          </button>
          <button
            aria-disabled={true}
            className="ms-Button ms-Button--icon is-disabled root-63"
            data-is-focusable={false}
            disabled={true}
            onClick={[Function]}
            onKeyDown={[Function]}
            onKeyPress={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            type="button"
          >
            <span
              className="ms-Button-flexContainer flexContainer-42"
              data-automationid="splitbuttonprimary"
            >
              <i
                aria-hidden={true}
                className="ms-Icon root-37 css-67 ms-Button-icon icon-64"
                data-icon-name="Down"
                role="presentation"
                style={
                  Object {
                    "fontFamily": "\\"FabricMDL2Icons-0\\"",
                  }
                }
              >
                
              </i>
            </span>
          </button>
          <button
            className="ms-Button ms-Button--icon root-68"
            data-is-focusable={true}
            onClick={[Function]}
            onKeyDown={[Function]}
            onKeyPress={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            type="button"
          >
            <span
              className="ms-Button-flexContainer flexContainer-42"
              data-automationid="splitbuttonprimary"
            >
              <i
                aria-hidden={true}
                className="ms-Icon root-37 css-49 ms-Button-icon icon-51"
                data-icon-name="Delete"
                role="presentation"
                style={
                  Object {
                    "fontFamily": "\\"FabricMDL2Icons\\"",
                  }
                }
              >
                
              </i>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div>
    <br />
    <div
      className="ms-Grid-col ms-sm12"
    >
      <button
        className="ms-Button ms-Button--primary root-50"
        data-is-focusable={true}
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        type="submit"
      >
        <span
          className="ms-Button-flexContainer flexContainer-42"
          data-automationid="splitbuttonprimary"
        >
          <span
            className="ms-Button-textContainer textContainer-43"
          >
            <span
              className="ms-Button-label label-52"
              id="id__31"
            >
              Submit
            </span>
          </span>
        </span>
      </button>
    </div>
  </div>
</form>
`;
